server/api/user/update/controller.test.js

```diff
@@ -0,0 +1,511 @@
+        },
+      });
+    });
+    test('returns error when specifying a primary email that is not verified', async () => {
+      const badEmails = [{
+        address: 'not@verified.com',
+        isPrimary: true,
+      }];
+      const res = await request(server)
+        .post('/api/user.update')
+        .set('Authorization', `Bearer ${session.accessToken}`)
+        .send({
+          id: user.id,
+          emails: badEmails,
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 10004,
+          message: `You need to verify ${badEmails[0].address} before marking it as primary`,
+        },
+      });
+    });
+    test('updates user and returns expected response', async () => {
+      const newPictureURL = 'https://www.acme.com/pictures/v2/coyote.png';
+      const res = await request(server)
+        .post('/api/user.update')
+        .set('Authorization', `Bearer ${session.accessToken}`)
+        .send({
+          id: user.id,
+          fullName: 'new Coyotee',
+          username: 'wile2',
+          emails: [{
+            address: 'coyote@acme.com',
+            isPrimary: true,
+          }, {
+            address: 'coyote-new@acme.com',
+          }],
+          picture: newPictureURL,
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: true,
+        user: {
+          id: user.id,
+          fullName: 'new Coyotee',
+          username: 'wile2',
+          emails: expect.arrayContaining([{
+            address: expect.any(String),
+            isVerified: expect.any(Boolean),
+            isPrimary: expect.any(Boolean),
+          }]),
+          picture: newPictureURL,
+        },
+      });
+    });
+    test('returns invalid password error if oldPassword is invalid', async () => {
+      const newPassword = 'thi$-$s-$af3r';
+      const res = await request(server)
+        .post('/api/user.update')
+        .set('Authorization', `Bearer ${session.accessToken}`)
+        .send({
+          id: user.id,
+          oldPassword: 'invalid-password',
+          password: newPassword,
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 10002,
+          message: expect.any(String),
+        },
+      });
+    });
+    test('updates user password', async () => {
+      const newPassword = 'thi$-$s-$af3r';
+      const res = await request(server)
+        .post('/api/user.update')
+        .set('Authorization', `Bearer ${session.accessToken}`)
+        .send({
+          id: user.id,
+          oldPassword: 'catch-the-b1rd$',
+          password: newPassword,
+        });
+
+      expect(res.status).toBe(200);
+      const newSession = await createSession(ctx, {
+        username: 'wile2', // this was changed in an earlier test. How can we avoid it best?
```

Comment:
Okay what do we do about this one? This username is dependant on the test prior to it changing. This means that this unit test is now going to fail in case the update test fails as well. This means we are keeping state and this is bad.

Both tests will fail at both times although, so it won't be _that_ much of a problem, but i still think its wrong. Should i switch the `username` property using the updateService after the test passes?

Reply:
Bear in mind we have similar issues with other tests as well.

The only solution I can think of is wrapping everything is `describe` and have `beforeAll` and `afterAll` create/destroy the test prerequisites. These methods run irrelevant to the test being successful or not.